src/clj/fluree/db/indexer/garbage.cljc

```diff
@@ -0,0 +1,91 @@
+            [fluree.db.util.async #?(:clj :refer :cljs :refer-macros) [<? go-try]]
+            [fluree.db.util.core :as util]
+            [fluree.db.util.log :as log]))
+
+#?(:clj (set! *warn-on-reflection* true))
+
+(defn garbage-meta-map
+  "Garbage file metadata that will be passed to the
+  garbage cleaning function that allows it to report
+  on status and have filename addresses."
+  [{:keys [prev-index garbage] :as idx-root}]
+  (when garbage ;; first index will not have garbage
+    (assoc garbage
+      ;; retain not the current index file, but the previous one
+      ;; where the garbage in this file originated
+      :t (:t prev-index)
+      :index (:address prev-index))))
+
+(defn trace-idx-roots
+  [conn commit]
+  (go-try
+   (loop [next-idx-root (<! (storage/read-db-root conn (-> commit :index :address)))
+          garbage       []]
+     (if (or (nil? next-idx-root) ;; no more idx-roots
+             (util/exception? next-idx-root)) ;; if idx-root already deleted, will be exception
+       garbage
+       (let [garbage-meta  (garbage-meta-map next-idx-root)
+             prev-idx-addr (get-in next-idx-root [:prev-index :address])
+             garbage*      (if garbage-meta
+                             (conj garbage garbage-meta)
+                             garbage)]
+         (recur (<! (storage/read-db-root conn prev-idx-addr))
+                garbage*))))))
+
+(defn clean-garbage-record
+  "Cleans up a complete garbage file, which will contain
+  many index segment garbage items within it. Garbage file
+  looks like:
+  {:alias 'ledger-alias',
+   :branch 'main',
+   :garbage ['fluree:file://.../index/segment/1.json', ...]
+   :t 1}"
+  [conn {:keys [address index t] :as _garbage-map}]
+  (go
+    (let [{:keys [alias branch garbage]} (<! (storage/read-garbage conn address))]
+      (log/info "Removing" (count garbage) "unused index segments (garbage) from ledger"
+                alias "branch" branch "from index-t of" t)
+
+      ;; first delete all index segment files
+      (doseq [garbage-item garbage]
+        ;; note if the file was already deleted there could be an exception.
+        ;; this might happen if the server shutdown in the middle of a garbage
+        ;; exceptions are logged downstream, so just swallow exception here
+        ;; and keep rocessing.
+        (<! (storage/delete-garbage-item conn garbage-item)))
+
+      ;; delete main garbage record
+      (<! (storage/delete-garbage-item conn address))
+
+      ;; then delete the parent index root
+      (<! (storage/delete-garbage-item conn index)))))
+
+(defn clean-garbage
+  "Cleans up garbage data for old indexes, but retains
+  the last `max-indexes` indexes. Note that if not retaining
+  at least one prior index, current queries or handles on the latest
+  db might fail, therefore `max-indexes` is recommended to typically
+  be 1 or more."
+  [{:keys [conn commit] :as _db} max-indexes]
+  (go-try
+   (assert (and (int? max-indexes)
```

Comment:
`assert` is something that makes me nervous because `AssertionError`s aren't part of the `Exception` tree, so it will bubble through all our exception handling and take down the process.

Reply:
At this point this is a background process to a background process (indexing) - there is nothing that is looking at the response at all, and nothing to capture any exceptions. If exceptions occur, it won't stop the operation of the ledger, but they will appear in the logs.

The main goal of the assert is to stop processing to avoid unexpected results, but not to produce anything that might use it upstream.

I'm happy to change this to an ex-info however - in case something does use the results of this some day it will be in a better place.

Reply:
Adjusted here: 20b919c2991b2ced63309270dab9f355186eec21